### ISSUE TITLE
External dialogs appear behind VSCode when launched from powershell

### ISSUE THREAD

**Issue by TylerLeonhardt** [microsoft/vscode#42356]
- VSCode Version: Code 1.19.2 (490ef761b76b3f3b3832eff7a588aac891e5fe80, 2018-01-10T15:55:03.538Z)
- OS Version: Windows_NT x64 10.0.16299
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
csharp|ms-|1.13.1
PowerShell|ms-|1.5.1

---

Steps to Reproduce:

With the AzureRM powershell module:

1. in the default powershell terminal run: `Login-AzureRmAccount`
2. Notice the dialog appear behind VSCode:
![image](https://user-images.githubusercontent.com/2644648/35529753-ded37ba2-04e6-11e8-8c34-323cbe6c111b.png)

Reproduces without extensions: Yes

I'm also curious if there's a way for an extension like the PowerShell extension for VSCode to bring UI to the top, over VSCode. Often UI generated by other programs (debug extension, user programs) show up behind VSCode and it confuses users (causes bug submissions).

This issue was sparked by:
https://github.com/PowerShell/vscode-powershell/issues/1175

**Comment by bpasero** [comment]
Reading your description it seems that the issue originates from an extension and not VS Code itself. Thereby I will close this issue and kindly ask you to report this issue against the extension repository. 

Most extensions are on GitHub as well so it should be easy to file it against the extension. If you open an extension on the market place, click on the **Support** link to find out more.

<img width="196" alt="screen shot 2016-05-05 at 12 51 17" src="https://cloud.githubusercontent.com/assets/900690/15041943/263e2282-12c0-11e6-9a33-d2b3b500d61b.png">

Thanks again for your feedback!

**Comment by argonym** [comment]
```
if (issue.description.match("extension") || issue.hasLabel("extension")) {
    issue.post(templates.refer_extension);
    issue.close();
}
```

**Comment by tylerl0706** [comment]
This was closed by mistake it would seem.

I stated in my repro steps:
"In the default powershell terminal". In other words, the one that ships with VSCode itself.

I was also just curious from an extension standpoint. That part is more of an aside to the actual issue.

**Comment by Tyriar** [comment]
Looks like this is related to electron, I haven't experienced the issue personally.

**Comment by tylerl0706** [comment]
@Tyriar here's an easier repro that doesn't require the Azure cmdlets.

```powershell
Get-Process | Out-GridView
```

Can you repro that?

**Comment by Tyriar** [comment]
Yes I can repro with that, thanks. I still think it's an Electron issue though

**Comment by chrmarti** [comment]
Some observations:
- Opening an external link from the release notes editor opens Edge in front.
- Opening the device login from the Azure Account extension opens Edge in front.
- Opening a link (`Start-Process "http://www.microsoft.com/"`) from PowerShell in the integrated terminal opens Edge behind.
- Opening a link (`Start-Process "http://www.microsoft.com/"`) from PowerShell in the external terminal opens Edge in front.

**Comment by corbob** [comment]
I have managed to find these Electron issues that sound related, but not completely sure:

https://github.com/electron/electron/issues/11989

https://github.com/electron/electron/issues/6081

**Comment by kowax** [comment]
Hi,
I found a lot of threads about the same issue, most of them is closed as related/duplicated issues but nothing what could explain/help or tell what to do and where is the problem. I have newly installed VS Code with powershell extension and observe same issue no matter if I run "login-azurermaccount" or "Get-Process | Out-GridView" all windows which should appear in new window are in the background. In Powershell ISE it works without any issues.

VS Code:
Version: 1.30.2 (user setup)
Commit: 61122f88f0bf01e2ac16bdb9e1bc4571755f5bd8
Date: 2019-01-07T22:54:13.295Z
Electron: 2.0.12
Chrome: 61.0.3163.100
Node.js: 8.9.3
V8: 6.1.534.41
OS: Windows_NT x64 10.0.17763

Extension:
powershell extension: 1.11.0

PC:
windows 10 ver. 1809 build 17763.253

Please, could somebody help?

Thank you.

**Comment by TylerLeonhardt** [comment]
> [looks like electron also passed on owning this issue](https://github.com/electron/electron/pull/10783#issuecomment-413599294) - they [refer to using `forcefocus` as a workaround](https://github.com/robinwassen/forcefocus). 
>
> It does seem to be a hard problem, but from a user experience perspective we could do better - maybe we can consider using `forcefocus` in vscode for this use case. If this is your first user experience with vscode like @cbailiss  - you are less likely to continue using the editor and shell as it doesn't work. 

Just wanted to pass along @alohaninja's comment in https://github.com/PowerShell/vscode-powershell/issues/621#issuecomment-479769873

As it seems like this will have to be addressed by vscode.

**Comment by poortom1004** [comment]
This is problematic for beginners on vscode.  I transitioned from the ISE to VS Code and experienced this.  I wasted 30 mins trying to figure it out.  Now multiply that amount of wasted time by the number of new VS Code users using the product simply for the PowerShell extension because Microsoft deprecated the PowerShell ISE.  That's a lot of wasted time.

**Comment by cbailiss** [comment]
I still find this very annoying every time I use VS Code - just bad user experience in the tool.
I appreciate that it is probably caused by issues spread across different teams/code at Microsoft, but I'm looking forward to the day someone at Microsoft finally takes ownership of it and deals with it.

**Comment by the-ress** [comment]
@Tyriar This is a ConPTY issue, not Electron: microsoft/terminal#2988

It’s the same underlying issue as in #929, #13255. The powershell process doesn’t have permission to change the foreground window.

Using [the-ress/node-windows-foreground-love](https://github.com/the-ress/node-windows-foreground-love) isn’t practical in this case - the permission gets revoked after next key press or mouse click.

Windows console [handles this by using undocumented API](https://github.com/microsoft/terminal/blob/4dd476ecbd2dca59f288304d2b7581c8ff4c6e38/src/server/ProcessList.cpp#L301), but there’s not an equivalent documented functionality for ConPTY (as far as I can tell).

**Comment by Tyriar** [comment]
@DustinCampbell any advice on what we should do? microsoft/terminal is using private APIs apparently to force launched apps to the foreground https://github.com/microsoft/vscode/issues/42356#issuecomment-552153955

**Comment by Tyriar** [comment]
Meant to ping @DHowett-MSFT

**Comment by DHowett-MSFT** [comment]
@Tyriar I'm somewhat affronted at your assertion that Terminal is using a private API we wouldn't expose to anybody else.

There's no magic here. It just luckily works out for Terminal sometimes. It's being tracked by https://github.com/microsoft/terminal/issues/2988, and there's really no good solution for us _or_ for you. PowerShell cmdlets need to not have a dependency on the window they're hosted in, because that will fail for all remoting scenarios and all SSH/telnet/multi-tenant scenarios too.

**Comment by DHowett-MSFT** [comment]
Reading up a few move comments:

@the-ress The snippet you identified [here](https://github.com/microsoft/terminal/blob/4dd476ecbd2dca59f288304d2b7581c8ff4c6e38/src/server/ProcessList.cpp#L295) serves a totally different purpose: it marks the application that spawned the console window as _owning_ the console window. This makes things like task manager and process explorer able to map from a HWND to _cmd.exe_ or _powershell.exe_.

That code isn't included in the Terminal application. :smile:

**Comment by TylerLeonhardt** [comment]
>  PowerShell cmdlets need to not have a dependency on the window they're hosted in, because that will fail for all remoting scenarios and all SSH/telnet/multi-tenant scenarios too.

Totally agree... but considering so many PowerShell customers have relied on:

* [Out-GridView](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/out-gridview?view=powershell-7) for visualizing data/manually filtering a pipeline
* WinForms & WPF (to a point where they've made tooling around it: https://poshgui.com/)
* Some Azure PowerShell cmdlets still pop a GUI... and [people run into it](https://twitter.com/DcrReynolds/status/1254777276756942850?s=20).

Unfortunately, this type of practice isn't going to stop any time soon.

**Comment by jeremyciak** [comment]
Any update on this? I am experiencing the same issue now myself with trying to implement a user-friendly `Out-GridView` step in a PowerShell driven process.

**Comment by JanVinterberg** [comment]
Is this dead or still being investigated?

**Comment by cbailiss** [comment]
Unless someone from Microsoft is able to confirm that they own this issue, it seems we must unfortunately conclude it is in the "too hard to solve" category / effectively dead.

**Comment by rzhao271** [comment]
Looks like the upstream issue's been fixed: https://github.com/microsoft/terminal/issues/2988.

**Comment by bmustari** [comment]
For what it's worth I can confirm that this issue of dialogs appearing behind VSCode is happening on Linux as well(Jammy 22.04).  This also seems to be unrelated to the terminal.  If I create a new .txt for instance(ctrl-n) and try to save it(ctrl-shift-s), the dialog for save-as is nowhere to be seen, assuming VSC is fullscreen.  I also use two monitors and thought perhaps it was related until I saw this issue here.  
To give a bit more info, this only started happening sometime within the last 2-3 updates to VSCode.  I was about to just revert to an earlier version since this is driving me mad, but I decided to search a bit and found this issue.
I have been happily using this editor for a few years now and it has been quite nice and productive, this dialog issue is enough to make me switch editors or revert to an old version and turn auto-updates off.  I really hope it can be fixed.
